### DISCUSSION TITLE
How to write to a vector aux variable and access the gradient of a vector aux variable?

### DISCUSSION THREAD

**Discussion by andrisfreimanis** [idaholab/moose#25751] [Q&A General] [answered]
Dear All,
I need to write a crystal plasticity material model that depends on the gradients of vectors defined for each slip system. Assuming 12 slip systems, a user would need to define 36 standard variables or 12 vector variables. Due to the smaller number I would like to use vector variables. 

From what I understand from Moose's documentation I could define vector variables and gradients in the header file as 
```
std::vector<VectorMooseVariable *> _g_vector_var;
std::vector<const VectorVariableGradient *> _g_vector_grad;
```
The `Coupleable` class documentation mentions and there are several examples in Moose using `Coupleable::writableVariable`. That returns a reference to `MooseWritableVariable` and I don't see an equivalent for `MooseVectorVariable.` What is the intended way of writing to vector variables?

Also, I can get the gradient pointer in the constructor with
```
for (auto const slipId : make_range(_number_slip_systems))
    _g_vector_grad[slipId] = &coupledVectorGradient("g_vector_variables", slipId);
```
Is this the intended way to access coupled vector gradients?
  
Best regards
Andris

**Comment by GiudGiud** [comment]
Hello

You are not supposed to write to variables using the Coupleable interface. This writable reference was a hack introduced to work with an external library. 

To write to a variable, you should get the solution from the system, get the dof index, then write that way.
The transfers do it for variable and array variables for an example.
For a vector variable it's pretty similar. 

Guillaume

**Reply by GiudGiud** [reply, marked as answer]
like this for example, in the projection transfer:
```
  for (const auto & node : to_mesh.local_node_ptr_range())
  {
    for (unsigned int comp = 0; comp < node->n_comp(to_sys.number(), to_var.number()); comp++)
    {
      const dof_id_type to_index = node->dof_number(to_sys.number(), to_var.number(), comp);
      to_solution->set(to_index, the value );
    }
```

**Comment by andrisfreimanis** [comment]
Thanks, that's the answer I was looking for.

Andris
